File: CSharpCodeBase/unity/entity.lua
local Class = require 'utils.hump.class'

local Entity = Class({init = function (self)
      self.components = {}
      self._awakened = false
end})

function Entity:AddComponent(componentType)
  self.components[componentType] = componentType(self)
  if self._awakened then self.components[componentType]:Awake() end
  return self.components[componentType]
end

function Entity:GetComponent(componentType)
  return self.components[componentType]
end

function Entity:Awake()
  for k, v in pairs(self.components) do v:Awake() end
  self._awakened = true
end

function Entity:Update() 
  for k, v in pairs(self.components) do v:Update() end
end

function Entity:FixedUpdate()
  for k, v in pairs(self.components) do v:FixedUpdate() end
end

function Entity:LateUpdate() 
  for k, v in pairs(self.components) do v:LateUpdate() end
end

function Entity:GetType()
  return "Entity"
end

function Entity:Message(funcname, data)
  for k, v in pairs(self.components) do
    func = v[funcname]
    if func ~= nil then
      func(v, data)
    end
  end
end

function Entity:Enable()
  self.enabled = true
end

function Entity:Disable()
  self.enabled = false
end

function Entity:Callback(method, ...)  
  if self[method] ~= nil then      
    local result = self[method](self, ...)
    return result
  else
    return {}  
  end
end

return Entity

File: CSharpCodeBase/unity/entity.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class entity {
 public void init (self);
       this.components = {}
       this._awakened = false;
 }})
 public void AddComponent(componentType){
   this.components[componentType] = componentType(self);
   if(this._awakened  ){  this.components[componentType]:Awake() }
   return this.components[componentType];
 }
 public void GetComponent(componentType){
   return this.components[componentType];
 }
 public void Awake(){
   for(k, v in pairs(this.components) ){ v:Awake() }
   this._awakened = true;
 }
 public void Update() {
   for(k, v in pairs(this.components) ){ v:Update() }
 }
 public void FixedUpdate(){
   for(k, v in pairs(this.components) ){ v:FixedUpdate() }
 }
 public void LateUpdate() {
   for(k, v in pairs(this.components) ){ v:LateUpdate() }
 }
 public void GetType(){
   return "Entity";
 }
 public void Message(funcname, data){
   for(k, v in pairs(this.components) ){
     func = v[funcname];
     if(func != null  ){ 
       func(v, data);
     }
   }
 }
 public void Enable(){
   this.enabled = true;
 }
 public void Disable(){
   this.enabled = false;
 }
 public void Callback(method, ...)  {
   if(self[method] != null  ){       
     var result = self[method](self, ...);
     return result;
   }else{
     return {}  
   }
 }
 }}